### PR TITLE
[SPARK-41349][CONNECT] Implement DataFrame.hint

### PR DESCRIPTION
**Pull request by dengziming**
### What changes were proposed in this pull request?

1. Implement `DataFrame.hint` for scala API
2. Implement `DataFrame.hint` for python API

### Why are the changes needed?

API coverage

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
UT

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @amaliujia FYI

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/LiteralValueProtoConverterSuite.scala [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+package org.apache.spark.sql.connect.planner
+
+import org.scalatest.funsuite.AnyFunSuite
+
+import org.apache.spark.sql.connect.planner.LiteralValueProtoConverter.{toCatalystValue, toConnectProtoValue}
+
+
+class LiteralValueProtoConverterSuite extends AnyFunSuite {
+
+  test("Round-trip") {
```
Comment (amaliujia) [inline connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/LiteralValueProtoConverterSuite.scala:27]: Nit: better replace this test name by `test proto value and catalyst value conversion`, optionally you can add the SPARK-xxxx at the beginning of the test name.

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectPlannerSuite.scala [hunk 2]
```diff
@@ -571,4 +572,28 @@ class SparkConnectPlannerSuite extends SparkFunSuite with SparkConnectPlanTest {
       Dataset.ofRows(spark, transform(proto.Relation.newBuilder.setProject(project).build()))
     assert(df.schema.fields.toSeq.map(_.name) == Seq("id"))
   }
+
+  test("Test Hint") {
```
Comment (amaliujia) [inline connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectPlannerSuite.scala:576]: Maybe also add a test case to have a random hint name? I expect Catalyst will throw something.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 3]
```diff
@@ -233,19 +235,9 @@ class SparkConnectPlanner(session: SparkSession) {
...
-        case proto.Expression.Literal.LiteralTypeCase.DOUBLE => value.getDouble
-        case proto.Expression.Literal.LiteralTypeCase.STRING => value.getString
-        case other => throw InvalidPlanInput(s"Unsupported value type: $other")
-      }
-    }
-
     val replacement = mutable.Map.empty[Any, Any]
     rel.getReplacementsList.asScala.foreach { replace =>
-      replacement.update(convert(replace.getOldValue), convert(replace.getNewValue))
+      replacement.update(toCatalystValue(replace.getOldValue), toCatalystValue(replace.getNewValue))
```
Comment (amaliujia) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:240]: In fact there is a `private def transformLiteral(lit: proto.Expression.Literal):` that covers all literal value conversion. 

Either re-use that one or merge that one with the `toCatalystValue`?

File: python/pyspark/sql/connect/plan.py [hunk 4]
```diff
@@ -32,6 +32,21 @@ class InputValidationError(Exception):
     pass
 
 
+def convert_value(v: Any) -> proto.Expression.Literal:
```
Comment (amaliujia) [inline python/pyspark/sql/connect/plan.py:35]: we have this: https://github.com/apache/spark/blob/master/python/pyspark/sql/connect/column.py#L178

is it feasible to extract the common part and re-use?

**Comment by amaliujia** [comment]
@dengziming there are several techniques that can help you:

when you working on Scala, you can run `./dev/lint-scala` locally which does scala side lint check. One of the check of that is the format over the Connect Scala module, which in fact you can run this command to auto-format the connect Scala code: `./build/mvn -Pscala-2.12 scalafmt:format -Dscalafmt.skip=fase -Dscalafmt.validateOnly=false -Dscalafmt.changedOnly=false -pl connector/connect`

When you working on Python side, after you build the Scala side jars, you can run `./python/run-tests --module pyspark-connect` to test your code (you may consider use https://docs.conda.io/en/latest/miniconda.html to manage the python envs and in the conda env you can `pip install -r dev/requirement.txt --upgrade`).

Similar, you can run `./dev/lint-python` to do the lint check. And there will be some prompted suggestions if you fail a check. 


Please feel free to ask questions that help you develop easier.

**Review by dengziming — COMMENTED** [review]
@amaliujia Thank you for your patience, I have fixed these comments, and I reverted the python related change since it's a little complex so I want to do it in another PR, PTAL.

**Review by amaliujia — COMMENTED** [review]
LGTM

It is ok to separate Python and Scala implementation.

**Comment by amaliujia** [comment]
@cloud-fan can you take a look?

**Comment by HyukjinKwon** [comment]
@dengziming mind rebasing this please? Otherwise, should be good to go.

**Comment by dengziming** [comment]
Hello @HyukjinKwon  , I saw the related changes has been reverted, so I also reverted my change.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala [hunk 5]
```diff
@@ -694,6 +685,18 @@ package object dsl {
           .build()
       }
 
+      def hint(name: String, parameters: Any*): Relation = {
+        Relation
+          .newBuilder()
+          .setHint(
+            Hint
```
Comment (cloud-fan) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:692]: hmm, the proto message `Hint` was added before this PR?

Comment (dengziming) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:692]: Yes, it was already added in #38857

**Review by cloud-fan — APPROVED** [review]

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/LiteralValueProtoConverter.scala [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+import org.apache.spark.sql.types._
+import org.apache.spark.unsafe.types.{CalendarInterval, UTF8String}
+
+object LiteralValueProtoConverter {
+
+  /**
+   * Transforms the protocol buffers literals into the appropriate Catalyst literal expression.
+   *
+   * @return
+   *   Expression
```
Comment (beliefer) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/LiteralValueProtoConverter.scala]:
```suggestion
   * @return Expression
```

Comment (dengziming) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/LiteralValueProtoConverter.scala]: Thank your for your suggestions, in fact, this is formatted by the scalafmt plugin, you can use `./dev/lint-scala` to format the code before submit a PR.

Comment (beliefer) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/LiteralValueProtoConverter.scala]: Yeah. But it seems strange. cc @amaliujia

Comment (amaliujia) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/LiteralValueProtoConverter.scala]: hmmm we actually enable scalafmt for Spark Connect code. So if this pass scalafmt that is fine.

Only Spark Connect enforce that so if you are seeing this is different from other modules, that is expected.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**